### PR TITLE
UX: pasting links on a selection will apply a link format

### PR DESCRIPTION
**Pull request by gwwar**
One handy way of adding a link is by copying a link url and pasting it into a text selection. This has been in use as a way of adding links in [WordPress](https://wordpress.org/) for quite some time.

If there isn't a text selection, or the copied item is not a link we should see default paste behavior. 

Changes in this PR implement the idea in Discourse as discussed in https://meta.discourse.org/t/idea-paste-links-in-the-editor

https://user-images.githubusercontent.com/1270189/143144443-01316a2b-5bd0-4742-8844-1b5f180012c6.mp4

### Testing Instructions

- No regressions for paste behavior in d-editor
- Verify that the new qunit tests make sense. In order to fully test paste events, we may need to switch these to integration tests, but let me know if folks have preferences.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/mixins/textarea-text-manipulation.js [hunk 1]
```diff
@@ -272,6 +273,19 @@ export default Mixin.create({
       }
     }
 
+    if (plainText && !handled && selected.end > selected.start) {
+      let isURL;
+      try {
+        isURL = !!new URL(plainText);
```
Comment (gwwar) [inline app/assets/javascripts/discourse/app/mixins/textarea-text-manipulation.js:279]: Let me know if folks have a preferred way of testing for links in Discourse. Generally it's link-like enough if the URL constructor is able to parse the string.

https://url.spec.whatwg.org/#valid-url-string

Comment (eviltrout) [reply app/assets/javascripts/discourse/app/mixins/textarea-text-manipulation.js:279]: This seems good to me, but I do wonder if we will end up linking things that are not links, and if that will be more of a nuissance than the thing the auto linking solves?

Comment (gwwar) [reply app/assets/javascripts/discourse/app/mixins/textarea-text-manipulation.js:279]: > I do wonder if we will end up linking things that are not links

In practice, I haven't seen a major report like this yet for folks using the block or classic editor in WordPress. We do get small issues from time to time, but they tend to be minor inconveniences.

That said, I browsed around the codebase some more and it does look like discourse-markdown uses https://github.com/markdown-it/linkify-it to test links. If we'd like this pasting behavior, ideally we can match the rules here. I'll take a look to see if it's accessible to d-editor.

Example of a cooked link in preview:
https://user-images.githubusercontent.com/1270189/142941831-7e07a6b5-8d10-4bb1-98c1-b723883f9c24.mp4

Another option might be to limit this to a site or user setting to toggle the behavior.

Comment (gwwar) [reply app/assets/javascripts/discourse/app/mixins/textarea-text-manipulation.js]: I sketched out an approach in 5a9cf8d where we reuse the linkify rules. Not sure if I like using library implementation details in this way, but I do feel like the overall behavior is much improved. One alternative might be using linkify directly versus off of the PrettyText instance.

 I'll add some example tests/videos tomorrow, but let me know if you had feedback/questions ahead of time.

**Review by eviltrout — COMMENTED** [review]
This is looking better. Strangely there is a test failure now for Ember CLI. Any ideas?

**Comment by gwwar** [comment]
> This is looking better. Strangely there is a test failure now for Ember CLI. Any ideas?

Thanks for taking another look. If this isn't a flakey test, the new init in the mixin could have affected timing for some tests. I'll try a few things.

So far this is green locally for me, but it's not uncommon to have CI be a bit more sensitive to timing issues.

For future me I verified this locally by:
1. Running the full suite from the `app/assets/javascripts` folder can check results via `~/checkouts/discourse/app/assets/javascripts/node_modules/.bin/ember test --launch Chrome`.
2. Running a single test: `~/checkouts/discourse/app/assets/javascripts/node_modules/.bin/ember test --launch Chrome --filter "Integration | Component | d-editor: bold with a multiline selection"`
3. Also via `./bin/rake qunit:test` from the root checkout directory.

Screen Shot 2021-11-23 at 1 58 37 PM

**Comment by gwwar** [comment]
@eviltrout I think the multiline test is flakey, if we visit https://github.com/discourse/discourse/actions/workflows/ember.yml?query=is%3Afailure and spot check a few items, it's also failing in other branches. For example:

https://github.com/discourse/discourse/runs/4298128519?check_suite_focus=true
https://github.com/discourse/discourse/runs/4262881361?check_suite_focus=true

**Review by eviltrout — APPROVED** [review]

**Comment by eviltrout** [comment]
Thanks! I've merged it. Let's give it a try.

**Comment by gwwar** [comment]
Nice! Thanks for the review @eviltrout!

**Comment by yaxu** [comment]
I've only just realised this much longed-for feature has been implemented, and it's made my day, thanks to @gwwar and all involved!

**Pull request merged**